src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.LazyItemOperation.cs

```diff
@@ -59,9 +59,24 @@ IList<I> GetReferencedItems(string itemType, ImmutableHashSet<string> globsToIgn
                 }
             }
 
-            public abstract void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore);
+            public virtual void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
+            {
+                var items = SelectItems(listBuilder, globsToIgnore).ToList();
+                MutateItems(items);
```

Comment:
This seems like it may modify the metadata on items from previous versions of the list, which are supposed to be immutable.  To avoid this, each item might need to be cloned before modifying the metadata.